pkg/mapper/dynamicfile/dynamicfile.go

```diff
@@ -0,0 +1,262 @@
+package dynamicfile
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/sirupsen/logrus"
+	"sigs.k8s.io/aws-iam-authenticator/pkg/config"
+)
+
+type DynamicFileMapStore struct {
+	mutex sync.RWMutex
+	users map[string]config.UserMapping
+	roles map[string]config.RoleMapping
+	// Used as set.
+	awsAccounts map[string]interface{}
+	filename    string
+}
+
+type DynamicFileData struct {
+	// RoleMappings is a list of mappings from AWS IAM Role to
+	// Kubernetes username + groups.
+	RoleMappings []config.RoleMapping `json:"mapRoles"`
+
+	// UserMappings is a list of mappings from AWS IAM User to
+	// Kubernetes username + groups.
+	UserMappings []config.UserMapping `json:"mapUsers"`
+	// AutoMappedAWSAccounts is a list of AWS accounts that are allowed without an explicit user/role mapping.
+	// IAM ARN from these accounts automatically maps to the Kubernetes username.
+	AutoMappedAWSAccounts []string `json:"mapAccounts"`
+}
+
+type ErrParsingMap struct {
+	errors []error
+}
+
+func (err ErrParsingMap) Error() string {
+	return fmt.Sprintf("error parsing dynamic file: %v", err.errors)
+}
+
+func waitUntilFileAvailable(filename string) error {
+	waitFileTimer := time.NewTimer(1 * time.Second)
```

Comment:
I agree that the timer is awkward here, because you have to reset it every iteration.  However, I think this code should be refactored.  I think we might want to consider using a separate goroutine which simple waits for the file, then sets up fsnotify.  If the file doesn't exist or is deleted, the goroutine can exit and be restarted by a ticker or it can handle that internally. This goroutine will handle everything in terms of fsnotify, reading the file, and unmarshalling it into a struct. 

The goroutine should be managed by something like `wait.Until()`, such as in `go wait.Until(c.runWorker, time.Second, ctx.Done())`.  This allows us to manage the goroutine with a context and to control how frequently the outermost loop runs.  

We can then either save the struct into the cache in this same goroutine, or if any additional processing needs to happen we can follow the pattern of adding a cache lookup key to a work queue for a worker to pick up.

Reply:
_This goroutine will handle everything in terms of fsnotify, reading the file, and unmarshalling it into a struct_ -- is this what we have here?

_We can then either save the struct into the cache in this same goroutine, or if any additional processing needs to happen we can follow the pattern of adding a cache lookup key to a work queue for a worker to pick up._. -- is this adding more complexity?

Reply:
The difference, which I did not clearly state, is that the goroutine should terminate and be restarted at an interval using `wait.Until().` It should also respect a context, and terminate when `context.Done()`.  1 goroutine might be fine in this case, but if additional complexity arises outside of reading from the file then I'd advocate for separating into 2 goroutines with distinct jobs.  Also, what about locking between this reader and the file writer?  I took a quick glance at fsnotify and it wasn't clear to me if we'd need the ability to use something like [flock](https://www.man7.org/linux/man-pages/man2/flock.2.html) too.

Reply:
if there is any pending change when the reader read the file, it is ok. When the pending change is completed, fsnotify will trigger a new WRITE event, reader will read the file again. If we put a lock there, then reader will be blocked till the writer completes. Then fsnotify still will trigger a WRITE event. So wondering flock is not necessary here